tests/unit/test_projects.py

```diff
@@ -1773,3 +1774,191 @@ def test_project_get_build_for_arch_triplet_all(self, project_yaml_data):
+
+        assert project.components
+        assert project.components["foo"].summary == summary
+
+    def test_component_summary_invalid(self, project_yaml_data, stub_component_data):
+        component = {"foo": stub_component_data}
+        component["foo"]["summary"] = "x" * 79
+
+        error = "ensure this value has at most 78 characters"
+        with pytest.raises(errors.ProjectValidationError, match=error):
+            Project.unmarshal(project_yaml_data(components=component))
+
+    @pytest.mark.parametrize(
+        "version",
+        [
+            "1",
+            "1.0",
+            "1.0.1-5.2~build0.20.04:1+1A",
+            "git",
+            "1~",
+            "1+",
+            "12345678901234567890123456789012",
+        ],
+    )
+    def test_component_version_valid(
+        self, version, project_yaml_data, stub_component_data
+    ):
+        component = {"foo": stub_component_data}
+        component["foo"]["version"] = version
+
+        project = Project.unmarshal(project_yaml_data(components=component))
+
+        assert project.components
+        assert project.components["foo"].version == version
+
+    @pytest.mark.parametrize(
+        "version,error",
+        [
+            (
+                "1_0",
+                "Invalid version '1_0': Component versions consist of",
+            ),  # _ is an invalid character
+            (
+                "1=1",
+                "Invalid version '1=1': Component versions consist of",
+            ),  # = is an invalid character
+            (
+                ".1",
+                "Invalid version '.1': Component versions consist of",
+            ),  # cannot start with period
+            (
+                ":1",
+                "Invalid version ':1': Component versions consist of",
+            ),  # cannot start with colon
+            (
+                "+1",
+                r"Invalid version '\+1': Component versions consist of",
+            ),  # cannot start with plus sign
+            # escaping + from the regex string to match.
+            (
+                "~1",
+                "Invalid version '~1': Component versions consist of",
+            ),  # cannot start with tilde
+            (
+                "-1",
+                "Invalid version '-1': Component versions consist of",
+            ),  # cannot start with hyphen
+            (
+                "1.",
+                "Invalid version '1.': Component versions consist of",
+            ),  # cannot end with period
+            (
+                "1:",
+                "Invalid version '1:': Component versions consist of",
+            ),  # cannot end with colon
+            (
+                "1-",
+                "Invalid version '1-': Component versions consist of",
+            ),  # cannot end with hyphen
+            (
+                "123456789012345678901234567890123",
+                "ensure this value has at most 32 characters",
+            ),  # too large
```

Comment:
I don't mind if this is convention, but the pytest-encouraged way to do this would be:

```suggestion
            pytest.param(
                "123456789012345678901234567890123",
                "ensure this value has at most 32 characters",
                id="too large",
            ),
```

Reply:
Maybe something like `"x" * 33` could be more fail-proof than a literal string.

Reply:
Thanks, fixed! (both here and where I copied from)